django/db/models/expressions.py

```diff
@@ -866,6 +895,112 @@ def as_sql(self, compiler, connection, template=None, case_joiner=None, **extra_
         return sql, sql_params
 
 
+class SubQuery(Expression):
+    """
+    An explicit subquery. It may contain OuterRef() references to the outer
+    query, which will be resolved when it is applied to that query.
+    """
+    template = '(%(subquery)s)'
+
+    def __init__(self, subquery, output_field=None, **extra):
+        self.subquery = subquery.all()
+        self.extra = extra
+        if output_field is None and len(self.subquery.query.select) == 1:
+            output_field = self.subquery.query.select[0].field
+        super(SubQuery, self).__init__(output_field)
+
+    def copy(self):
+        clone = super(SubQuery, self).copy()
+        # Also create a new copy of the subquery.
+        clone.subquery = clone.subquery.all()
+        return clone
+
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
+        clone = self.copy()
+        clone.is_summary = summarize
+        clone.subquery.query.bump_prefix(query)
+
+        # Need to recursively resolve these.
+        def resolve_all(child):
+            if hasattr(child, 'children'):
+                [resolve_all(_child) for _child in child.children]
+            if hasattr(child, 'rhs'):
+                child.rhs = resolve(child.rhs)
+
+        def resolve(child):
+            if hasattr(child, 'resolve_expression'):
+                return child.resolve_expression(
+                    query=query, allow_joins=allow_joins, reuse=reuse, summarize=summarize,
+                    for_save=for_save)
+            return child
+
+        resolve_all(clone.subquery.query.where)
+
+        for key, value in clone.subquery.query.annotations.items():
+            if isinstance(value, SubQuery):
+                clone.subquery.query.annotations[key] = resolve(value)
+
+        return clone
+
+    def get_source_expressions(self):
+        return [x for x in [getattr(expr, 'lhs', None) for expr in self.subquery.query.where.children] if x]
+
+    def relabeled_clone(self, change_map):
```

Comment:
@mjtamlyn I used this instead of `set_source_expressions`, because otherwise it turns `self.subquery` into a list, which breaks other stuff. I believe it has the desired effect though (of updating the source expressions).